src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -124,6 +128,129 @@ private static String generateUserInputCode() {
         return sb.toString();
     }
 
+    public static ConsumeCodeResponse consumeCode(Main main, String deviceId, String userInputCode, String linkCode)
+            throws RestartFlowException, ExpiredUserInputCodeException, IncorrectUserInputCodeException,
+            StorageTransactionLogicException, StorageQueryException, NoSuchAlgorithmException, InvalidKeyException {
+        PasswordlessSQLStorage passwordlessStorage = StorageLayer.getPasswordlessStorage(main);
+        long passwordlessCodeLifetime = Config.getConfig(main).getPasswordlessCodeLifetime();
+        int maxCodeInputAttempts = Config.getConfig(main).getPasswordlessMaxCodeInputAttempts();
+
+        String deviceIdHash;
+        String linkCodeHash;
+        if (linkCode != null) {
+            byte[] linkCodeBytes = Base64.getUrlDecoder().decode(linkCode);
+            linkCodeHash = Utils.hashSHA256Base64(linkCodeBytes);
+
+            PasswordlessCode code = passwordlessStorage.getCodeByLinkCodeHash(linkCodeHash);
+            if (code == null || code.createdAt < (System.currentTimeMillis() - passwordlessCodeLifetime)) {
+                throw new RestartFlowException();
+            }
+            deviceIdHash = code.deviceIdHash;
+        } else {
+            byte[] deviceIdBytes = Base64.getDecoder().decode(deviceId);
+            deviceIdHash = Utils.hashSHA256Base64UrlSafe(deviceIdBytes);
+
+            byte[] linkCodeBytes = Utils.hmacSHA256(deviceIdBytes, userInputCode);
+            linkCodeHash = Utils.hashSHA256Base64(linkCodeBytes);
+        }
+
+        PasswordlessDevice consumedDevice;
+        try {
+            consumedDevice = passwordlessStorage.startTransaction(con -> {
+                PasswordlessDevice device = passwordlessStorage.getDevice_Transaction(con, deviceIdHash);
+                if (device == null) {
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
+                if (device.failedAttempts >= maxCodeInputAttempts) {
+                    passwordlessStorage.deleteDevice_Transaction(con, deviceIdHash);
+                    passwordlessStorage.commitTransaction(con);
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
```

Comment:
This `if` would never actually be true cause of the `if` in line 176. So I assume that this is here cause we want to take into account the case where the dev might reduce the value of `maxCodeInputAttempts`. Is this true? If so, please add a comment here explaining this - as it's not very obvious.